src/draive/splitters/basic.py

```diff
@@ -33,74 +45,71 @@ def split_text(
+            if next_index < len(splitters):
+                remained_splitters = list(splitters[next_index:])
+            else:
+                remained_splitters = None
 
-    # try splitting using provided splitter
-    parts: list[str] = text.split(splitter)
-    # if splitting has done nothing retry using secondary splitter
-    if len(parts) == 1:
-        splitter = alt_splitter
-        parts = text.split(splitter)
+            return (splitter, remained_splitters, parts)
     # if splitting has still done nothing then fail
-    if len(parts) == 1:
-        raise ValueError("Failed to properly split text with provided separators")
+    raise ValueError("Failed to properly split text with provided separators")
 
-    return (splitter, parts)
 
-
-def _merge(  # noqa: C901
+def _merge(  # noqa: C901, PLR0912, PLR0913
     parts: list[str],
     part_size: int,
     count_size: Callable[[str], int],
     splitter: str,
+    remained_splitters: list[str] | None,
     part_overlap_size: int | None,
 ) -> list[str]:
     result: list[str] = []
     accumulator: list[str] = []
     accumulator_size: int = 0
     # iterate over splitted pats
     for part in parts:
-        temp_size: int = count_size(part)
+        current_part: str = part
+        # Add the separator back if it's not the first part
+        if accumulator:
+            current_part = splitter + part
+
+        temp_size: int = count_size(current_part)
         # check if can add to previous part
-        if accumulator_size + temp_size < part_size:
-            accumulator.append(part)
+        if accumulator_size + temp_size <= part_size:
+            accumulator.append(current_part)
             accumulator_size += temp_size
         # check if part is not too big on its own
         elif temp_size > part_size:
-            chunk: str = splitter.join(accumulator).strip()
+            chunk: str = "".join(accumulator).strip()
             if chunk:
                 result.append(chunk)
             # do special splitting if it is too big indeed
             # force overlap and split on newlines and spaces
             result.extend(
                 split_text(
-                    text=part,
+                    text=current_part,
                     part_size=part_size,
-                    separators=("\n", " "),
-                    part_overlap_size=part_overlap_size
-                    or int(part_size * 0.2),  # if there was no overlap force at least 20%
+                    separators=remained_splitters,
+                    part_overlap_size=part_overlap_size,
                     count_size=count_size,
                 ),
             )
             accumulator = []
             accumulator_size = 0
         # if we have overlap defined do overlap between last part and current (not fitting)
         elif part_overlap_size := part_overlap_size:
-            chunk: str = splitter.join(accumulator).strip()
+            chunk: str = "".join(accumulator).strip()
```

Comment:
this seems to be a bug - anyways we should probably keep accumulator as a string

Reply:
ahh ok, I have forgot about the overlap 🤔